catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/service/impl/TransformerLocatorImpl.java

```diff
@@ -0,0 +1,218 @@
+        MetacardTransformer.class,
+        MetacardTransformerAdapter::new,
+        filter);
+  }
+
+  @Override
+  public List<QueryResponseTransformer> findQueryResponseTransformers(MimeType mimeType) {
+    return new ArrayList<>(
+        filterByMimeType(
+            findTransformers(
+                QueryResponseTransformer.class, QueryResponseTransformerAdapter::new, null),
+            mimeType));
+  }
+
+  @Override
+  public List<QueryResponseTransformer> findQueryResponseTransformersByFilter(String filter) {
+    return findTransformers(
+        QueryResponseTransformer.class, (transformer, properties) -> transformer, filter);
+  }
+
+  private <T extends TransformerProperties> List<T> filterByMimeType(
+      List<T> transformers, MimeType userMimeType) {
+
+    String userIdValue = userMimeType.getParameter(Constants.SERVICE_ID);
+
+    String userBaseType = userMimeType.getBaseType();
+
+    return transformers
+        .stream()
+        .filter(s -> CollectionUtils.isNotEmpty(s.getMimeTypes()))
+        .filter(transformer -> getBaseTypes(transformer).contains(userBaseType))
+        .filter(
+            transformer ->
+                userIdValue == null
+                    || StringUtils.equals(userIdValue, getServiceId(transformer.getProperties())))
+        .collect(Collectors.toList());
+  }
+
+  private Set<String> getBaseTypes(TransformerProperties transformerProperties) {
+    return transformerProperties
+        .getMimeTypes()
+        .stream()
+        .map(MimeType::getBaseType)
+        .collect(Collectors.toSet());
+  }
+
+  private String getServiceId(Map<String, Object> properties) {
+    Object idServiceProperty = properties.get(Constants.SERVICE_ID);
+
+    if (idServiceProperty != null) {
+
+      return idServiceProperty.toString();
+    }
+
+    return null;
+  }
+
+  private Map<String, Object> getServiceProperties(ServiceReference serviceReference) {
+    return Arrays.stream(serviceReference.getPropertyKeys())
+        .collect(Collectors.toMap(s -> s, serviceReference::getProperty));
+  }
+
+  protected BundleContext getContext() {
+    Bundle cxfBundle = FrameworkUtil.getBundle(TransformerLocatorImpl.class);
+    if (cxfBundle != null) {
+      return cxfBundle.getBundleContext();
+    }
+    return null;
+  }
+
+  private <T> Collection<ServiceReference<T>> findServices(Class<T> clazz, String filter) {
+
+    BundleContext bundleContext = getContext();
+
+    try {
+      return bundleContext.getServiceReferences(clazz, filter);
+    } catch (InvalidSyntaxException e) {
+      throw new IllegalArgumentException("Invalid syntax supplied");
+    }
+  }
+
+  private <T, R> List<R> findTransformers(
+      Class<T> clazz, BiFunction<T, Map<String, Object>, R> converter, String filter) {
+    BundleContext bundleContext = getContext();
+
+    Collection<ServiceReference<T>> refs = findServices(clazz, filter);
+
+    if (refs != null) {
+      List<ServiceReference<T>> list = new ArrayList<>(refs);
```

Comment:
Also, I think this could be done without an extra list:
```
return refs.stream()
  .sorted(reverseOrder())
  .map(...)
  .collect(toList());
```

Reply:
done